### ISSUE TITLE
Refactoring Collection API

### ISSUE THREAD

**Issue by lintool** [castorini/anserini#667]
Looking at the current structure of the Collection API, I'm a bit puzzled as to why it was designed this way... perhaps it's accumulated debt?

It seems like the conceptual organization is as follows: A collection is comprised of segments. Each segment is backed by a file and provides an iterator of documents.

So, this definition makes complete sense:

```
public interface Segment<T extends SourceDocument> extends Iterator<T>
```

But why doesn't `DocumentCollection` extend an iterator over Segments, or at the very least have a `getSegments` method? Instead, there's this `SegmentProvider` class, and a collection, like `ClueWeb09Collection`, has the following definition:

```
public class ClueWeb09Collection extends DocumentCollection
    implements SegmentProvider<ClueWeb09Collection.Document>
```

Why? Why implement an interface as oppose to extend an abstract class?

@emmileaf can you take a look at this and think through what the Collection API _should_ be? While we're developing the Python bridge, we might as well see if we need to clean up the API on the Java end.

This should probably happen post v0.5.0 release...

**Comment by emmileaf** [comment]
`SegmentProvider` offers something similar to `getSegments`, represented as a list of paths:
https://github.com/castorini/anserini/blob/7fd1f648dcdd196cda436bb6af129c015dbdb9b6/src/main/java/io/anserini/collection/SegmentProvider.java#L29-L38

I agree that these methods can move to the `DocumentCollection` abstract class instead, getting rid of the interface. However, from #331  and #332  it seems like the original intent of having the `SegmentProvider` interface was to make the `collection-segment-document` relationship more explicit in signatures. Do we still want this to be the case?

**Comment by lintool** [comment]
I agree we should make the collection/segment/document relationship explicit, but I don't think the current design does that... for example, in theory we could have a collection without a provider... and classes other than collections can implement the provider... which may be confusing.

IIRC, we originally considered the possibility of other types of segments... segments backed by a file, segments backed by a stream (e.g., live Twitter feed), etc. We were probably overly ambitious, as non-file-backed segments never materialized... so `FileSegment` and `Segment` can probably be collapsed.

Now that I think about it, I remember why we designed it this way... `Iterator` is an interface, so a collection would need to be something like:

```
public interface DocumentCollection<D> extends Iterator<Segment<D>>
```

But in Java, interfaces can't have associated code (unlike Scala)... A proposed solution would be to move the `discover` method in the current `DocumentCollection` into a helper class, e.g., `PathCollector`. That way `DocumentCollection` can remain an interface.

I think this also solves the problem that you referenced above: potentially too many files open... This is why having a `List<Segment<T>> getSegments()` method is potentially dangerous, but having an iterator is slightly better (although a programmer could still iterate through all the segments, collect all the references... but such a strange usage would hopefully be caught in CR).

So, tl;dr - how about:

```
public interface DocumentCollection<T extends SourceDocument> extends Iterable<Segment<T>>
public interface Segment<T extends SourceDocument> extends Iterable<T>
public class PathCollector
```

Also, note I changed from `Iterator` to `Iterable` - the distinction is subtle but important. A segment is *not* itself an iterator - it is a collection that can be iterated over (and same w/ document collection).

Thoughts? @emmileaf @r-clancy

**Comment by r-clancy** [comment]
As of Java 8, interfaces can have `static` and `default` methods (see [here](https://www.baeldung.com/java-static-default-methods)). Could the `discover` method be implemented as such in `DocumentCollection` to get rid of the extra helper class?

Otherwise, makes sense to me.

**Comment by emmileaf** [comment]
Makes sense. Just to check I've understood correctly... right now we have
- `DocumentCollection` contains `FileSegments` which are iterators over `SourceDocuments`

and after refactoring, we would have:  
- `DocumentCollection` is an iterable of `FileSegments`, which are iterables of `SourceDocuments`

Since FileSegment is currently implemented as an iterator, we'd be wrapping it into something like this?

```
public interface FileSegment<T extends SourceDocument> extends Iterable<T>{
	Iterator<T> iterator(); // collection-specific iterator logic moved to here 
}
```

**Comment by lintool** [comment]
I think that's right, but without actually rolling up yourselves to do the refactoring, I don't know if there are any hidden landmines...

Here's my suggestion - revamp the subclasses, and do one impl (e.g., NYT or WaPo) - send a preliminary PR - we can take a look and give feedback... then it's just retrofitting the other collections?

**Comment by emmileaf** [comment]
Sounds good, I'll give that a try first! 

An additional thought: with python/java bridge in mind, what kind of document content (raw, parsed) do we want to make available through the collection API? 

1. Currently `collection` and `generator` have some overlapping roles in the document processing pipeline. For example, parsing and selection of what attributes to index is done in [collection](https://github.com/castorini/anserini/blob/7fd1f648dcdd196cda436bb6af129c015dbdb9b6/src/main/java/io/anserini/collection/NewYorkTimesCollection.java#L1866-L1868) for NYT, but  [generator](https://github.com/castorini/anserini/blob/7fd1f648dcdd196cda436bb6af129c015dbdb9b6/src/main/java/io/anserini/index/generator/WapoGenerator.java#L94-L143) in WashPost.

2. If we set up python bridge to collection for raw text and generator for parsing, `generator` depends on `IndexCollection.Args` to know what fields to store in the lucene index, and `IndexCollection.Counters` to keep track of errors/empty files. The current python wrapper setup instantiates these, and gets a bit messy. 

I think it'd be helpful to decouple either of these on the Java end. Thoughts on two possible options?

__Option 1:__ the current python wrapper setup, but decoupling `generator`/`IndexCollection`
- collection API -> provide `SourceDocument` (raw contents only) -> 
generator API -> provide `LuceneDocument` (`FIELD_BODY` contains parsed main contents)

__Option 2:__ move some parsing logic to collection classes, decoupling `collection`/`generator`
- collections API -> provide `SourceDocument` (raw contents + parsed main contents), or
- collections API -> provide `MultiFieldSourceDocument` (raw contents and parsed attributes)

**Comment by lintool** [comment]
Let's punt on this decision for now? Ideally, the collection API should only be responsible for procuring `SourceDocument`s... we can circle back and revamp the parsing/selection/to Lucene conversion as part of a "data cleaning" stage?

But I think this is closer to Option 2 above?
